// moonlink.js/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Connectors = exports.WebSocket = exports.PlayerManager = exports.NodeManager = exports.SearchResult = exports.DeezerSource = exports.SpotifySource = exports.YouTubeLiveChat = exports.VoiceReceiver = exports.Track = exports.Filters = exports.Rest = exports.Queue = exports.Player = exports.Node = exports.Manager = exports.version = void 0;
exports.version = require("../package.json").version;
const Util_1 = require("./src/Util");
const Manager_1 = require("./src/core/Manager");
Object.defineProperty(exports, "Manager", { enumerable: true, get: function () { return Manager_1.Manager; } });
const Node_1 = require("./src/entities/Node");
Object.defineProperty(exports, "Node", { enumerable: true, get: function () { return Node_1.Node; } });
const Player_1 = require("./src/entities/Player");
Object.defineProperty(exports, "Player", { enumerable: true, get: function () { return Player_1.Player; } });
const Queue_1 = require("./src/entities/Queue");
Object.defineProperty(exports, "Queue", { enumerable: true, get: function () { return Queue_1.Queue; } });
const Rest_1 = require("./src/entities/Rest");
Object.defineProperty(exports, "Rest", { enumerable: true, get: function () { return Rest_1.Rest; } });
const Filters_1 = require("./src/entities/Filters");
Object.defineProperty(exports, "Filters", { enumerable: true, get: function () { return Filters_1.Filters; } });
const Track_1 = require("./src/entities/Track");
Object.defineProperty(exports, "Track", { enumerable: true, get: function () { return Track_1.Track; } });
const VoiceReceiver_1 = require("./src/entities/VoiceReceiver");
Object.defineProperty(exports, "VoiceReceiver", { enumerable: true, get: function () { return VoiceReceiver_1.VoiceReceiver; } });
const YouTubeLiveChat_1 = require("./src/entities/YouTubeLiveChat");
Object.defineProperty(exports, "YouTubeLiveChat", { enumerable: true, get: function () { return YouTubeLiveChat_1.YouTubeLiveChat; } });
const SearchResult_1 = require("./src/structures/SearchResult");
Object.defineProperty(exports, "SearchResult", { enumerable: true, get: function () { return SearchResult_1.SearchResult; } });
const NodeManager_1 = require("./src/managers/NodeManager");
Object.defineProperty(exports, "NodeManager", { enumerable: true, get: function () { return NodeManager_1.NodeManager; } });
const PlayerManager_1 = require("./src/managers/PlayerManager");
Object.defineProperty(exports, "PlayerManager", { enumerable: true, get: function () { return PlayerManager_1.PlayerManager; } });
const WebSocket_1 = require("./src/services/WebSocket");
Object.defineProperty(exports, "WebSocket", { enumerable: true, get: function () { return WebSocket_1.WebSocket; } });
const connectors_1 = require("./src/connectors");
Object.defineProperty(exports, "Connectors", { enumerable: true, get: function () { return connectors_1.Connectors; } });
const spotify_1 = require("./src/sources/spotify");
Object.defineProperty(exports, "SpotifySource", { enumerable: true, get: function () { return spotify_1.SpotifySource; } });
const deezer_1 = require("./src/sources/deezer");
Object.defineProperty(exports, "DeezerSource", { enumerable: true, get: function () { return deezer_1.DeezerSource; } });
Util_1.Structure.register("Node", Node_1.Node);
Util_1.Structure.register("Player", Player_1.Player);
Util_1.Structure.register("Queue", Queue_1.Queue);
Util_1.Structure.register("Rest", Rest_1.Rest);
Util_1.Structure.register("Filters", Filters_1.Filters);
Util_1.Structure.register("Track", Track_1.Track);
Util_1.Structure.register("SearchResult", SearchResult_1.SearchResult);
Util_1.Structure.register("NodeManager", NodeManager_1.NodeManager);
Util_1.Structure.register("PlayerManager", PlayerManager_1.PlayerManager);
Util_1.Structure.register("WebSocket", WebSocket_1.WebSocket);
